### PR TITLE
[Merged by Bors] - chore: remove some no-longer-needed backwards compatibility flags

### LINKED ISSUE #12535: tracking issue for `set_option backward.isDefEq.lazyProjDelta false`
This option disables the changes to `isDefEq` introduced in https://github.com/leanprover/lean4/pull/3965, and is still sometimes necessary in Mathlib.

It would be great to reduce our reliance on this backwards compatibility flag.

This may involve
* just reducing the reliance on defeq in the affected theorems (i.e. more rewrites, less rfl/convert/...)
* producing minimized examples of the places in Mathlib where the old behaviour is required for performance.

(See [all occurrences on Mathlib's `master`](https://github.com/search?q=repo%3Aleanprover-community%2Fmathlib4+backward.isDefEq.lazyProjDelta&type=code).)

### PR DESCRIPTION
**Pull request by kim-em**

### CODE REVIEW

**Comment by jcommelin** [comment]
!bench

File: Mathlib/Algebra/Order/GroupWithZero/Canonical.lean [hunk 1]
```diff
@@ -225,15 +225,10 @@ theorem OrderIso.mulRight₀'_symm {a : α} (ha : a ≠ 0) :
...
-#adaptation_note /-- 2024-04-23
-After https://github.com/leanprover/lean4/pull/3965,
-we need to either write `@inv_zero (G₀ := α) (_)` in `neg_top`,
-or use `set_option backward.isDefEq.lazyProjDelta false`.
-See https://github.com/leanprover-community/mathlib4/issues/12535 -/
 instance : LinearOrderedAddCommGroupWithTop (Additive αᵒᵈ) :=
   { Additive.subNegMonoid, instLinearOrderedAddCommMonoidWithTopAdditiveOrderDual,
     Additive.instNontrivial with
-    neg_top := set_option backward.isDefEq.lazyProjDelta false in @inv_zero _ (_)
+    neg_top := inv_zero (G₀ := α)
```
Comment (Vierkantor) [inline Mathlib/Algebra/Order/GroupWithZero/Canonical.lean:231]: I definitely agree with this change.

**Comment by Vierkantor** [comment]
I was about to suggest the same thing :)

**Comment by kim-em** [comment]
!bench

**Review by Vierkantor — APPROVED** [review]
I agree that a 7-10% slowdown of the whole file is enough to warrant keeping the options enabled. Thank you for adding more explanation to the `set_option`s!

Like this looks good to me, feel free to merge if the benchmarking passes:

bors d+

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**